### PR TITLE
[TypeSpecRequirement] Convert to GitHub Action

### LINKED ISSUE #28861: Migrate TypeSpecRequirement to GitHub Action

### PR DESCRIPTION
**Pull request by mikeharder**
- Fixes #28861
- Test Build: https://github.com/Azure/azure-rest-api-specs/pull/29627

### CODE REVIEW

File: eng/scripts/Logging-Functions.ps1 [hunk 1]
```diff
@@ -10,6 +14,9 @@ function LogWarning {
   if (Test-SupportsDevOpsLogging) {
     Write-Host ("##vso[task.LogIssue type=warning;]$args" -replace "`n", "%0D%0A")
   }
+  elseif (Test-SupportsGitHubLogging) {
```
Comment (konrad-jamrozik) [inline eng/scripts/Logging-Functions.ps1:17]: I guess we could dedup all these `if/else` blocks into a function that smartly formats the message to fit ADO or GH?

File: eng/pipelines/typespec-requirement.yml [hunk 2]
```diff
@@ -1,22 +0,0 @@
-trigger: none
```
Comment (konrad-jamrozik) [inline eng/pipelines/typespec-requirement.yml:1]: Just for my learning: how the automation know to find and execute this file? Like, I know github looks inside `.github/workflows` by convention. Does something similar exist for GH/ADO integration, or do we have some config for that, or what?

**Review by konrad-jamrozik — APPROVED** [review]

File: .github/workflows/typespec-requirement.yaml [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+      fail-fast: false
+      matrix:
+        spec-type: [data-plane, resource-manager]
+
+    runs-on: ubuntu-latest
+
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          fetch-depth: 2
```
Comment (weshaggard) [inline .github/workflows/typespec-requirement.yaml:17]: Why depth 2?

File: eng/scripts/Logging-Functions.ps1 [hunk 4]
```diff
@@ -20,6 +27,9 @@ function LogErrorForFile($file, $errorString)
   if (Test-SupportsDevOpsLogging) {
     Write-Host ("##vso[task.logissue type=error;sourcepath=$file;linenumber=1;columnnumber=1;]$errorString" -replace "`n", "%0D%0A")
   }
+  elseif (Test-SupportsGitHubLogging) {
+    Write-Host ("::error file=$file,line=1,col=1::$errorString" -replace "`n", "%0D%0A")
```
Comment (weshaggard) [inline eng/scripts/Logging-Functions.ps1:31]: It is cool that GH actions support these logging commands as well. They look to support a few more we might want to take advantage of at some point as well like the summary https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#example-of-adding-a-job-summary.

File: .github/workflows/typespec-requirement.yaml [hunk 5]
```diff
@@ -0,0 +1,32 @@
+name: TypeSpec Requirement
+
+on: pull_request
+
+jobs:
+  TypeSpec-Requirement:
```
Comment (weshaggard) [inline .github/workflows/typespec-requirement.yaml:6]:
```suggestion
  TypeSpec-Requirement:
    name: TypeSpec Requirement
```

You can control the name of the job, including the space by specifying the name property under the job.

File: eng/pipelines/typespec-requirement.yml (hunk 2 above)
Comment (weshaggard) [reply eng/pipelines/typespec-requirement.yml:1]: That is configured on the DevOps side and stored with the pipeline you setup.

File: .github/workflows/typespec-requirement.yaml (hunk 5 above)
Comment (mikeharder) [reply .github/workflows/typespec-requirement.yaml:6]: If this works, we might not need to update openapi-alps at all!

File: eng/scripts/Logging-Functions.ps1 (hunk 4 above)
Comment (mikeharder) [reply eng/scripts/Logging-Functions.ps1:31]: I agree the ability to generate a custom job summary by writing markdown to a file on disk (via an env var) is very powerful.

File: .github/workflows/typespec-requirement.yaml (hunk 3 above)
Comment (mikeharder) [reply .github/workflows/typespec-requirement.yaml:18]: Default is depth 1 for perf.  However, TypeSpecRequirement needs to compute the `git diff` against `HEAD^`, which requires depth of at least two (current commit and previous commit).

Comment (mikeharder) [reply .github/workflows/typespec-requirement.yaml:19]: I added comments and parameters making this more clear

File: eng/pipelines/typespec-requirement.yml (hunk 2 above)
Comment (mikeharder) [reply eng/pipelines/typespec-requirement.yml:1]: After this file is deleted, the DevOps pipeline will be kept for history (and maybe older branches), and eventually deleted

File: eng/scripts/Logging-Functions.ps1 (hunk 1 above)
Comment (mikeharder) [reply eng/scripts/Logging-Functions.ps1:17]: I agree the code could be improved a bit by refactoring, and won't object if you want to submit a PR, but I didn't think it was worth it.  It's still a pretty small file, and I don't see it growing much more.

**Pull request merged**